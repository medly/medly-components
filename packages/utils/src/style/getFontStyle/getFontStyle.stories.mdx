import { Meta } from '@storybook/addon-docs/blocks';

<Meta title="Utils/Style/getFontStyle" />

# getFontStyle

The `getFontStyle` is a  styled component helper that generates font-related styles based on the passed theme. It accepts a specific font-variant or font-weight to influence the generated output.

### Example

The simple usage without any arguments must fall back on using configured defaults for font-related styles.

```
const plainParagraph = styled('p')`
    ${getFontStyle}
`;
```

Any specified font-variant or font-weight must take precedence in applying.

```
const seriousParagraph = styled('p')`
    ${({ theme }) => getFontStyle({ theme, fontWeight: 'ExtraStrong' })}
`;
```
