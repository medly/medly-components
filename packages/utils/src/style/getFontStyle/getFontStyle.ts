import { FontVariants, FontWeights, Theme } from '@medly-components/theme';
import { css } from 'styled-components';

export const getFontStyle = ({
    fontVariant,
    fontWeight,
    theme
}: {
    theme: Theme;
    fontVariant?: FontVariants;
    fontWeight?: FontWeights;
}) => {
    const { variants, weights, defaults } = theme.font,
        { fontSize, fontWeight: defaultFontWeight, letterSpacing, lineHeight } = variants[fontVariant || defaults.variant];

    return css`
        font-size: ${fontSize};
        font-weight: ${weights[fontWeight || defaultFontWeight]};
        letter-spacing: ${letterSpacing};
        line-height: ${lineHeight};
        font-family: ${defaults.fontFamily};
    `;
};
