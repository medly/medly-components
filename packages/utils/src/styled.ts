import * as styledComponents from 'styled-components';

const { default: styled, css, createGlobalStyle, keyframes, ThemeProvider, withTheme } = styledComponents;

export { styled, css, createGlobalStyle, keyframes, ThemeProvider, withTheme };
