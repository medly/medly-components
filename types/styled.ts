import 'styled-components';
import { Theme } from '../packages/theme';

declare module 'styled-components' {
    export interface DefaultTheme extends Theme {}
}
