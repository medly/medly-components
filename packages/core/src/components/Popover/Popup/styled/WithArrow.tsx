import { css } from 'styled-components';
import { PopoverPopupProps } from '../types';

const getPositionWithArrow = ({ placement }: PopoverPopupProps) => {
    switch (placement) {
        case 'top':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateX(-50%);
                    top: 100%;
                    left: 50%;
                }
            `;
        case 'top-start':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    top: 100%;
                    left: 2rem;
                }
            `;
        case 'top-end':
            return css`
                box-shadow: 0 0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    top: 100%;
                    right: 2rem;
                }
            `;
        case 'right':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateY(-50%) rotate(90deg);
                    top: 50%;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'right-start':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(90deg);
                    top: 2rem;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'right-end':
            return css`
                box-shadow: -0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(90deg);
                    bottom: 2rem;
                    right: calc(100% - 0.3rem);
                }
            `;
        case 'bottom':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateX(-50%) rotate(180deg);
                    left: 50%;
                    bottom: 100%;
                }
            `;
        case 'bottom-start':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(180deg);
                    left: 2rem;
                    bottom: 100%;
                }
            `;
        case 'bottom-end':
            return css`
                box-shadow: 0 -0.4rem 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(180deg);
                    right: 2rem;
                    bottom: 100%;
                }
            `;
        case 'left':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: translateY(-50%) rotate(270deg);
                    top: 50%;
                    left: calc(100% - 0.3rem);
                }
            `;
        case 'left-start':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(270deg);
                    top: 2rem;
                    left: calc(100% - 0.3rem);
                }
            `;
        case 'left-end':
            return css`
                box-shadow: 0.4rem 0 1.6rem rgba(176, 188, 200, 0.5);
                &::after {
                    transform: rotate(270deg);
                    bottom: 2rem;
                    left: calc(100% - 0.3rem);
                }
            `;
    }
};

export const getStyleWithArrow = css<PopoverPopupProps>`
    border-radius: 0.8rem;
    padding: 1.5rem 2rem;

    &::after {
        content: '';
        width: 0;
        height: 0;
        z-index: 4;
        position: absolute;
        border-left: 0.8rem solid transparent;
        border-right: 0.8rem solid transparent;
        border-top: 1rem solid ${({ bgColor }) => bgColor};
    }

    ${getPositionWithArrow}
`;
