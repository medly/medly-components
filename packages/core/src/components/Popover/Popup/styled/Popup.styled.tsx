import { fullHeight, fullWidth } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import { Placement } from '../../types';
import { StyledPopoverPopupProps } from '../types';
import { getStyleWithArrow } from './WithArrow';

export const getPosition = ({ placement, distance = '0px' }: { placement: Placement; distance?: string }) => {
    switch (placement) {
        case 'top-start':
            return css`
                bottom: calc(100% + ${distance});
                left: 0%;
            `;
        case 'top':
            return css`
                bottom: calc(100% + ${distance});
                left: 50%;
                transform: translateX(-50%);
            `;
        case 'top-end':
            return css`
                bottom: calc(100% + ${distance});
                right: 0%;
            `;
        case 'right-start':
            return css`
                top: 0%;
                left: calc(100% + ${distance});
            `;
        case 'right':
            return css`
                left: calc(100% + ${distance});
                top: 50%;
                transform: translateY(-50%);
            `;
        case 'right-end':
            return css`
                bottom: 0;
                left: calc(100% + ${distance});
            `;
        case 'bottom-start':
            return css`
                top: calc(100% + ${distance});
                left: 0%;
            `;
        case 'bottom':
            return css`
                top: calc(100% + ${distance});
                left: 50%;
                transform: translateX(-50%);
            `;
        case 'bottom-end':
            return css`
                top: calc(100% + ${distance});
                right: 0%;
            `;
        case 'left-start':
            return css`
                right: calc(100% + ${distance});
                top: 0%;
            `;
        case 'left':
            return css`
                right: calc(100% + ${distance});
                top: 50%;
                transform: translateY(-50%);
            `;
        case 'left-end':
            return css`
                right: calc(100% + ${distance});
                bottom: 0%;
            `;
    }
};

export const PopupStyled = styled('div')<StyledPopoverPopupProps>`
    z-index: 4;
    position: absolute;
    background-color: ${({ bgColor }) => bgColor};

    ${props => props.fullWidth && fullWidth()};
    ${props => props.fullHeight && fullHeight()};

    ${props => getPosition({ placement: props.placement, distance: props.withArrow ? '1.5rem' : props.distance })}
    ${({ withArrow }) => withArrow && getStyleWithArrow}
`;
