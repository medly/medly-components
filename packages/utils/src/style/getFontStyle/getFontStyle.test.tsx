import { render } from '@test-utils';
import styled from 'styled-components';
import { getFontStyle } from './getFontStyle';

describe('getFontStyle', () => {
    it('should apply default styles', () => {
        const Element = styled('div')`
            ${getFontStyle}
        `;
        const { container } = render(<Element />);
        expect(container).toMatchSnapshot();
    });

    it('should apply styles for specific font variant', () => {
        const Element = styled('div')`
            ${({ theme }) => getFontStyle({ theme, fontVariant: 'h1' })}
        `;
        const { container } = render(<Element />);
        expect(container).toMatchSnapshot();
    });

    it('should apply styles for a specific font weight', () => {
        const Element = styled('div')`
            ${({ theme }) => getFontStyle({ theme, fontWeight: 'ExtraStrong' })}
        `;
        const { container } = render(<Element />);
        expect(container).toMatchSnapshot();
    });

    it('should apply styles for a specific font variant and font weight', () => {
        const Element = styled('div')`
            ${({ theme }) => getFontStyle({ theme, fontVariant: 'h1', fontWeight: 'ExtraStrong' })}
        `;
        const { container } = render(<Element />);
        expect(container).toMatchSnapshot();
    });
});
