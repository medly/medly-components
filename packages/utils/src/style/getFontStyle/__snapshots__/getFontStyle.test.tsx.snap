// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFontStyle should apply default styles 1`] = `
.c0 {
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`getFontStyle should apply styles for a specific font variant and font weight 1`] = `
.c0 {
  font-size: 4.8rem;
  font-weight: 800;
  -webkit-letter-spacing: -0.04rem;
  -moz-letter-spacing: -0.04rem;
  -ms-letter-spacing: -0.04rem;
  letter-spacing: -0.04rem;
  line-height: 5.6rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`getFontStyle should apply styles for a specific font weight 1`] = `
.c0 {
  font-size: 1.4rem;
  font-weight: 800;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`getFontStyle should apply styles for specific font variant 1`] = `
.c0 {
  font-size: 4.8rem;
  font-weight: 600;
  -webkit-letter-spacing: -0.04rem;
  -moz-letter-spacing: -0.04rem;
  -ms-letter-spacing: -0.04rem;
  letter-spacing: -0.04rem;
  line-height: 5.6rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

<div>
  <div
    class="c0"
  />
</div>
`;
